feat: #878 - added a multiselect mode to product list page

monsieurtanuki opened the pull request:
Impacted files:
* `personalized_ranking_page.dart`: added a `fromItems` constructor for more flexibility
* `product_list_item_simple.dart`: added optional `onTap` parameter in order to override the default behavior when checking/unchecking
* `product_list_page.dart`: added a "multiselect" mode
* `product_query_page.dart`: refactored
* `scan_page_helper.dart`: refactored
* `smooth_it_model.dart`: for more flexibility, now using `List<Product>` instead of `ProductList`
* `smooth_product_card_found.dart`: added optional `onTap` parameter in order to override the default behavior when checking/unchecking

- solves openfoodfacts/smooth-app#878 

I'm 99% sure that you won't like the UI, but I couldn't think of another UX...

Product list page, with a new multi-select FAB:
![Simulator Screen Shot - iPhone 8 Plus - 2022-01-28 at 10 17 22](https://user-images.githubusercontent.com/11576431/151520096-fb2a1608-39ec-44d5-9d30-114a77c98092.png)

Multi-select mode, with a ranking FAB that goes to the ranking page with the selected items only:
![Simulator Screen Shot - iPhone 8 Plus - 2022-01-28 at 10 18 38](https://user-images.githubusercontent.com/11576431/151520324-46ad67e7-174f-48ef-b92e-5eebe52e536c.png)

monsieurtanuki committed:
Merge pull request #20 from openfoodfacts/develop

wtf

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
Merge branch 'openfoodfacts:develop' into develop

monsieurtanuki committed:
feat: #878 - added a multiselect mode to product list page

Impacted files:
* `personalized_ranking_page.dart`: added a `fromItems` constructor for more flexibility
* `product_list_item_simple.dart`: added optional `onTap` parameter in order to override the default behavior when checking/unchecking
* `product_list_page.dart`: added a "multiselect" mode
* `product_query_page.dart`: refactored
* `scan_page_helper.dart`: refactored
* `smooth_it_model.dart`: for flexibility, now using `List<Product>` instead of `ProductList`
* `smooth_product_card_found.dart`: added optional `onTap` parameter in order to override the default behavior when checking/unchecking

teolemon commented:
@monsieurtanuki just curious, why not placing text buttons above as outlined in the mocks ?
![image](https://user-images.githubusercontent.com/1689815/151526391-1a53fe92-8536-477f-85df-7ba0cdf232ae.png)

monsieurtanuki commented:
@teolemon The mocks do not reflect the current state of the app, and as I mentioned earlier currently there's an additional FAB on the bottom of the page.
Putting on top a button to compare, and on bottom a button to display the personalized ranking page, looked a bit confusing. Especially if the consequence is that you can select/unselect the products you compare, but not the products for the ranking page.
Unless the compare and ranking page are the same thing. I assumed that "compare" meant a new feature with products side by side, like when you compare smartphones memory, screensize, ... Or like openfoodfacts/openfoodfacts-server#6584.

teolemon commented:
Ah sorry for the confusion, the are absolutely the same thing. 
Rationale: most of the time, it does not make sense to compare all your history (comparing Carrots and Apples). So the compare button replaces the Personal Ranking, but does the same thing (trigger a personal ranking on the item selected).
Again sorry if that behaviour was not clear.

monsieurtanuki committed:
feat: #878 - added header with buttons; removed FABs

Impacted files:
* `product_list_page.dart`: added header with buttons; removed FABs
* `ranking_floating_action_button.dart`: added a comment about getting rid of this class because of possible confusion

monsieurtanuki committed:
feat: #878 - one cancel button is enough

Impacted files:
* `product_list_page.dart`: added header with buttons; removed FABs

monsieurtanuki commented:
I've just pushed this:
![Simulator Screen Shot - iPhone 8 Plus - 2022-01-28 at 11 47 54](https://user-images.githubusercontent.com/11576431/151533930-1fad9a3e-0afa-44a5-b0f2-67f537bada56.png)

![Simulator Screen Shot - iPhone 8 Plus - 2022-01-28 at 11 50 51](https://user-images.githubusercontent.com/11576431/151534339-64c11e97-0832-4745-b6eb-bbc2b803ccb8.png)

It's more like what you had in mind I guess, only less beautiful :)

jasmeet0817 commented on packages/smooth_app/lib/pages/product/common/product_list_page.dart:
```diff
@@ -48,75 +49,159 @@ class _ProductListPageState extends State<ProductListPage> {
+          if (dismissible && !_selectionMode)
+            PopupMenuButton<String>(
+              itemBuilder: (final BuildContext context) =>
+                  <PopupMenuEntry<String>>[
```
instead of String, can we use an Enum here?

Then the switch case can be exhaustive, no need for default: throw exception

jasmeet0817 commented on packages/smooth_app/lib/pages/product/common/product_list_page.dart:
```diff
@@ -48,75 +49,159 @@ class _ProductListPageState extends State<ProductListPage> {
+                      if (_selectionMode && _selectedBarcodes.isNotEmpty)
+                        ElevatedButton(
+                          child: const Text(
+                              'Compare'), // TODO(monsieurtanuki): localize
```
nit: If this goes in a second line, then add a comma after 'Compare' for flutter format to fix the formatting.

jasmeet0817 commented on packages/smooth_app/lib/pages/product/common/product_list_page.dart:
```diff
@@ -48,75 +49,159 @@ class _ProductListPageState extends State<ProductListPage> {
+                                    PersonalizedRankingPage.fromItems(
+                                  products: list,
+                                  title:
+                                      '', // TODO(monsieurtanuki): find inspiration
```
haha. How about "Your ranking" (was decided in a meeting with Stephane and Pierre)

jasmeet0817 commented on packages/smooth_app/lib/pages/product/common/product_list_page.dart:
```diff
@@ -48,75 +49,159 @@ class _ProductListPageState extends State<ProductListPage> {
+                            },
+                          ),
+                          child: const Text(
+                              'Compare Mode'), // TODO(monsieurtanuki): localize
```
optional: why not just localize in this PR :D

monsieurtanuki replied:
Good idea, but that would mean the labels are not supposed to change.
And I don't think it's the case as we're still discussing the UI/UX, e.g. if we finally put `'14/16 products'`, and where.

jasmeet0817 commented on packages/smooth_app/lib/pages/product/common/product_list_page.dart:
```diff
@@ -48,75 +49,159 @@ class _ProductListPageState extends State<ProductListPage> {
+                        ElevatedButton(
+                          onPressed: () =>
+                              setState(() => _selectionMode = false),
+                          child: Text(AppLocalizations.of(context)!.cancel),
```
maybe create appLocalization variable and use it ? We will need it in other places (when Todos are done).

jasmeet0817 commented:
Thank you @monsieurtanuki the code looks great, but question on the UI.

1. Do we need the AppBar in this page? There's no back button and in the mocks the title "History" appears in the header on the page, on the left of "Compare" button. Although "14/16 products selected" is perhaps useful (but I can't make up my mind on it). But in the mocks the title on the left and compare button on the right definitely looks better than just compare button in the center.
2. What do the two selector buttons do? I'm guessing they are used for select all or select none (which is great), but why is one of them checked already (when not everything is selected) and can't we live with just one ? (I guess this also ties into (1), if we want to have this checkbox, where will it go without AppBar). @teolemon any comment ?
![image](https://user-images.githubusercontent.com/7541730/151559784-65844a70-5cfa-4e8d-a648-bbfd0863894a.png)

teolemon commented:
This could be useful info, but:
- inline instead of the topbar
- "Select all" - "Select none" instead of the checkboxes (icons are too ambiguous)
- The number of products could be in the compare button ?

monsieurtanuki committed:
feat: #878 - closer to the mocks

Impacted file:
* `product_list_page.dart`: closer to the mocks

monsieurtanuki commented:
My point was to reproduce a standard behavior seen on standard apps, as I don't see the added value of reinventing the wheel.
For instance, you seem to be pretty relaxed about using two labels ("History" and "Compare Mode") in buttons on the same line. That mean we should hope that in no language both are long. For the record in French "History" is spelled "Mes produits récemment vus"... And instead of "History", in general the product list name is expected. Better be short.

Anyway, I've just pushed something close to the mocks:

![Simulator Screen Shot - iPhone 8 Plus - 2022-01-28 at 17 12 06](https://user-images.githubusercontent.com/11576431/151581848-06bc7b93-7638-4773-9d75-2785dd696319.png)

![Simulator Screen Shot - iPhone 8 Plus - 2022-01-28 at 17 12 31](https://user-images.githubusercontent.com/11576431/151581912-a3302776-8819-4275-8efd-c4c10b00d4b7.png)

I'm in favor of more standard (probably old-fashioned) behaviors - just a matter of taste I guess:
![Screenshot_2022-01-28-16-34-48](https://user-images.githubusercontent.com/11576431/151577631-da64f9b7-cd40-4aa1-8964-c1a8711512c3.png)

monsieurtanuki committed:
feat: #878 - unrelated minor refactoring

Impacted file:
* `product_query.dart`: unrelated removal of a harmless duplicate

teolemon commented:
@monsieurtanuki perhaps we can try to replace the title by the history icon ?

monsieurtanuki commented:
> @monsieurtanuki perhaps we can try to replace the title by the history icon ?

Yes we can but we'll still have the problem with other product list titles.
I'll send you a screenshot this afternoon.

monsieurtanuki commented:
![Simulator Screen Shot - iPhone 8 Plus - 2022-01-30 at 14 26 52](https://user-images.githubusercontent.com/11576431/151701697-c0abd5bb-b17a-46cd-8300-acee388195b5.png)

M123-dev commented on packages/smooth_app/lib/pages/product/common/product_list_page.dart:
```diff
@@ -48,62 +48,62 @@ class _ProductListPageState extends State<ProductListPage> {
+            if (_selectionMode && _selectedBarcodes.isNotEmpty)
+              ElevatedButton(
+                child: Text(
+                    'Compare ${_selectedBarcodes.length} products'), // TODO(monsieurtanuki): localize
```
I think we already have one for that (The button in the scanner)

M123-dev commented:
Just tested this PR and I like the feature to choose which products to compare. The code looks good and I don't want to really join the UI/UX discussion but I have a few minor points, nothing which has to be changed immediately:


- The "Cancel" button switched position after selecting the first product is selected
- I am also in favour of keeping some standard behavour (I think we already talked about this a while ago @monsieurtanuki)  and I was going to long tap on one of the products out of habit to turn on the compare mode, that would be one thing which we could maybe add.
- We should probably rename "My recently seen products" to "History" as in the mocks since it's not readable

monsieurtanuki commented:
Thank you @M123-dev for your comments.

> * The "Cancel" button switched position after selecting the first product is selected

Fair enough, I need to fix this.

> * I am also in favour of keeping some standard behavour (I think we already talked about this a while ago @monsieurtanuki)  and I was going to long tap on one of the products out of habit to turn on the compare mode, that would be one thing which we could maybe add.

I'm OK with that too.

> * We should probably rename "My recently seen products" to "History" as in the mocks since it's not readable

As evoked previously, this is not really the problem of the "history" title, it's more the problem of the additional text button on the right. We'll have the same problem with all product lists if they have long titles.
That said, I'm not in charge of the translations and I have no idea how to control the impact to all translations after a mere change in one label in English.
But feel free to rename it.

jasmeet0817 commented:
> Thank you @M123-dev for your comments.
> 
> > * The "Cancel" button switched position after selecting the first product is selected
> 
> Fair enough, I need to fix this.
> 
> > * I am also in favour of keeping some standard behavour (I think we already talked about this a while ago @monsieurtanuki)  and I was going to long tap on one of the products out of habit to turn on the compare mode, that would be one thing which we could maybe add.
> 
> I'm OK with that too.
> 
> > * We should probably rename "My recently seen products" to "History" as in the mocks since it's not readable
> 
> As evoked previously, this is not really the problem of the "history" title, it's more the problem of the additional text button on the right. We'll have the same problem with all product lists if they have long titles. That said, I'm not in charge of the translations and I have no idea how to control the impact to all translations after a mere change in one label in English. But feel free to rename it.

Yeah translations in such small real estate is always a problem, lets check this code in, and continue to think about how we can handle this.

Agreed on all other points.

More broader concern I have: 
We should also think about what really is "History". After scanning a few products I habitually tried to find them in this tab, but only products that I opened appear here, which was not obvious.

monsieurtanuki committed:
feat: #878 - UX fine-tuning

Impacted files:
* `app_en.arb`: added a "compare mode" label
* `app_fr.arb`: added a "compare mode" label
* `product_list_item_simple.dart`: added a `onLongPress` parameter
* `product_list_page.dart`: added longPress-to-selection-mode feature; fixed the cancel button position; localized

monsieurtanuki commented:
Thank you @teolemon @jasmeet0817 @M123-dev for your feedback!
* The size of the page title that can still be problematic
* I miss my "select all" / "select none" buttons, but I couldn't find where to put them
* I fixed the position of the "cancel" button
* I implemented the long press to selection mode
* it's localized
* probably won't look good on dark mode

monsieurtanuki commented:
> We should also think about what really is "History". After scanning a few products I habitually tried to find them in this tab, but only products that I opened appear here, which was not obvious.

@jasmeet0817 it's just the way it worked so far, I'm not saying it's a good idea. Besides, I cannot scan, therefore I'm not very sensitive about that.
But that would definitely make sense to add each scanned product to the history. Please create an issue for that, if relevant.
